compiler/grainlsp/completion.re

```diff
@@ -0,0 +1,136 @@
+open Grain;
+open Compile;
+open Grain_parsing;
+open Grain_utils;
+open Grain_typed;
+
+// maps Grain types to LSP CompletionItemKind
+let rec get_kind = (desc: Types.type_desc) =>
+  switch (desc) {
+  | TTyVar(_) => Utils.item_kind_completion_variable
+  | TTyArrow(_) => Utils.item_kind_completion_function
+  | TTyTuple(_) => Utils.item_kind_completion_struct
+  | TTyRecord(_) => Utils.item_kind_completion_struct
+  | TTyConstr(_) => Utils.item_kind_completion_constructor
+  | TTySubst(s) => get_kind(s.desc)
+  | TTyLink(t) => get_kind(t.desc)
+  | _ => Utils.item_kind_completion_text
+  };
+
+let process_resolution =
+    (
+      ~id,
+      ~compiled_code: Stdlib__hashtbl.t(string, Typedtree.typed_program),
+      ~cached_code: Stdlib__hashtbl.t(string, Typedtree.typed_program),
+      ~documents,
+      request,
+    ) => {
+  // right now we just resolve nothing to clear the client's request
+  // In future we may want to send more details back with Graindoc details for example
+  Rpc.send_completion(
+    ~output=stdout,
+    ~id,
+    [],
+  );
+};
+
+let process_completion =
+    (
+      ~id,
+      ~compiled_code: Stdlib__hashtbl.t(string, Typedtree.typed_program),
+      ~cached_code: Stdlib__hashtbl.t(string, Typedtree.typed_program),
+      ~documents,
+      request,
+    ) => {
+  switch (Utils.get_text_document_uri_and_position(request)) {
+  | (Some(uri), Some(line), Some(char)) =>
+    let completable = Utils.get_original_text(documents, uri, line, char);
+    switch (completable) {
+    | Nothing => Rpc.send_completion(stdout, id, [])
+    | Lident(text) =>
+      if (!Hashtbl.mem(cached_code, uri)) {
+        // no compiled code available
+        Rpc.send_completion(stdout, id, []);
+      } else {
+        let compiledCode = Hashtbl.find(cached_code, uri);
+        let modules = Env.get_all_modules(compiledCode.env);
+        let firstChar = text.[0];
```

Comment:
It won't throw as it's wrapped in a test for membership above as I really hate exceptions.
But as this could lead to problems down the line if someone changed something, I'm changing it to switch on the exception instead

Reply:
But I love camelcase :-D. Fixed.

Reply:
Or even just use the safe call that returns optional, even better.